licensing/notice.go

```diff
@@ -0,0 +1,283 @@
+}
+
+// GenerateNotice inspects the `go.mod` file and inspects the contents of the
+// downloaded dependency to determine which license is most predominant in its
+// source files. When a Writer is passed in the parameters, it also writes a
+// templated output in the writer.
+func GenerateNotice(params GenerateNoticeParams) (Notice, error) {
+	var notice Notice
+	if err := params.Validate(); err != nil {
+		return notice, err
+	}
+
+	params = params.fillDefaults()
+	notice = buildNotice(params)
+
+	paths, err := getModulePaths(params.GoModFile)
+	if err != nil {
+		return notice, err
+	}
+
+	notice.Dependencies = getLicenses(params.AnalyseFunc, paths...)
+
+	// Skip template execution.
+	if params.Writer == nil {
+		return notice, nil
+	}
+
+	templateFormat := fmt.Sprintf(NoticeFormat, params.NoticeHeader)
+	if err := writeTemplate(&notice, templateFormat, params.Writer); err != nil {
+		return notice, err
+	}
+	return notice, nil
+}
+
+// buildNotice constructs a Notice from the parameters.
+func buildNotice(params GenerateNoticeParams) (notice Notice) {
+	var currentYear = time.Now().Year()
+	notice = Notice{
+		Licensor:     params.Licensor,
+		Project:      params.Project,
+		ProjectYears: fmt.Sprint(params.StartYear, "-", currentYear),
+	}
+	if params.StartYear == currentYear || params.StartYear == 0 {
+		notice.ProjectYears = fmt.Sprint(currentYear)
+	}
+	return notice
+}
+
+// getModulePaths opens and parses a `go.mod` file obtaining the list of
+// dependencies and constructing a local filesystem path where they have
+// been downloaded. The returning slice is sorted.
+func getModulePaths(file string) ([]string, error) {
+	contents, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	module, err := gomod.Parse(file, contents)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(module.Require) == 0 {
+		return nil, errors.New("modfile has no dependencies to generate notice")
+	}
+
+	var paths = make([]string, 0, len(module.Require))
+	for dep, version := range module.Require {
+		re, err := regexp.Compile("([A-Z]+)")
+		if err != nil {
+			return nil, err
+		}
+
+		for _, letter := range re.FindAllString(dep, -1) {
+			dep = strings.Replace(dep, letter,
+				"!"+strings.ToLower(letter), -1,
+			)
+		}
+
+		p := filepath.Join(
+			append([]string{goPkgPath()}, strings.Split(dep, "/")...)...,
+		)
+		p = fmt.Sprint(p, "@", version)
+		paths = append(paths, p)
+	}
+
+	sort.Strings(paths)
+	return paths, nil
+}
+
+// getLicenses returns a []Dependency which is populated by the analyseFunc.
+// It relies on the local downloaded dependencies in order for the license
+// discovery mechanism to work. The Dependencies slice is sorted by License.
+func getLicenses(analyseFunc func(args ...string) []licensedb.Result, paths ...string) []Dependency {
+	var dependencies = make([]Dependency, 0, len(paths))
+	for _, res := range analyseFunc(paths...) {
+		arg := strings.Replace(res.Arg, goPkgPath()+"/", "", -1)
```

Comment:
Any chance to get something a bit more readable here?
I really have hard times to follow the code in the next 10-12 lines :(
Maybe adding some comments or extracting pieces to private functions with meaningful names?

Reply:
I've added comments here